barretenberg/cpp/src/barretenberg/vm2/simulation/bytecode_manager.test.cpp

```diff
@@ -0,0 +1,157 @@
+#include "barretenberg/vm2/simulation/bytecode_manager.hpp"
+
+#include <cstdint>
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <memory>
+#include <optional>
+#include <vector>
+
+#include "barretenberg/vm2/common/aztec_types.hpp"
+#include "barretenberg/vm2/common/field.hpp"
+#include "barretenberg/vm2/simulation/bytecode_hashing.hpp"
+#include "barretenberg/vm2/simulation/events/bytecode_events.hpp"
+#include "barretenberg/vm2/simulation/events/event_emitter.hpp"
+#include "barretenberg/vm2/simulation/testing/mock_dbs.hpp"
+#include "barretenberg/vm2/simulation/testing/mock_poseidon2.hpp"
+#include "barretenberg/vm2/simulation/testing/mock_range_check.hpp"
+#include "barretenberg/vm2/testing/fixtures.hpp"
+
+using ::testing::_;
+using ::testing::Return;
+using ::testing::SizeIs;
+using ::testing::StrictMock;
+
+namespace bb::avm2::simulation {
+
+namespace {
+
+// Simple mock for ContractInstanceManagerInterface
+class MockContractInstanceManager : public ContractInstanceManagerInterface {
+  public:
+    MOCK_METHOD(std::optional<ContractInstance>, get_contract_instance, (const FF& contract_address), (override));
+};
+
+class BytecodeManagerTest : public ::testing::Test {
+  protected:
+    BytecodeManagerTest()
+        : bytecode_hasher(poseidon2, hashing_events)
+    {}
+
+    StrictMock<MockContractDB> contract_db;
+    StrictMock<MockHighLevelMerkleDB> merkle_db;
+    StrictMock<MockPoseidon2> poseidon2;
+    StrictMock<MockRangeCheck> range_check;
+    StrictMock<MockContractInstanceManager> contract_instance_manager;
+
+    EventEmitter<BytecodeRetrievalEvent> retrieval_events;
+    EventEmitter<BytecodeDecompositionEvent> decomposition_events;
+    EventEmitter<InstructionFetchingEvent> instruction_fetching_events;
+    EventEmitter<BytecodeHashingEvent> hashing_events;
+    BytecodeHasher bytecode_hasher;
+};
+
+TEST_F(BytecodeManagerTest, RetrievalAndDeduplication)
+{
+    TxBytecodeManager tx_bytecode_manager(contract_db,
+                                          merkle_db,
+                                          poseidon2,
+                                          bytecode_hasher,
+                                          range_check,
+                                          contract_instance_manager,
+                                          retrieval_events,
+                                          decomposition_events,
+                                          instruction_fetching_events);
+
+    // Setup for base case
+    AztecAddress address1 = AztecAddress::random_element();
+    ContractInstance instance1 = testing::random_contract_instance();
+    ContractClass klass = testing::random_contract_class();
+
+    // Expected interactions for first retrieval
+    EXPECT_CALL(contract_instance_manager, get_contract_instance(address1))
+        .WillOnce(Return(std::optional<ContractInstance>(instance1)));
```

Comment:
Optional: consider `std::make_optional` (I think then you can avoid the `<>`) same in others.